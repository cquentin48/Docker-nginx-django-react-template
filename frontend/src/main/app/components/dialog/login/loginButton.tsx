import React from "react";
import { useLoginMutation } from "../../../store/user/userSlice";
import { type APIResponse } from "../../../model/user/httpRequestInterfaces";
import localizedStrings from "../../../model/locale/locale";
import { Button } from "@mui/material";
import { VpnKey } from "@mui/icons-material";
import { type VariantType } from "notistack";
import { ButtonDialogSX } from "../styles";
import UserFactory from "../../../model/user/userFactory";
import MessageNotification from "../../../model/message/message";

interface LoginButtonProps {
    username: string
    password: string
    displaySnackBar: (message: MessageNotification) => void
    handleClose: () => void
}

export const LoginButton = (props: LoginButtonProps): JSX.Element => {
    /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
    const [login, { isLoading }] = useLoginMutation();

    const handleLoginButton = async (props: LoginButtonProps): Promise<void> => {
        const loginInput = { username: props.username, password: props.password }
        const result = await login(loginInput).unwrap()

        let message;
        let severity: VariantType;
        if ('detail' in result) {
            message = localizedStrings.FAILED_LOGIN
            severity = "error"
        } else {
            const data: APIResponse = result;
            UserFactory.updateUser(data.access);
            message = localizedStrings.SUCCESSFUL_LOGIN
            severity = "success"
        }

        const notification = new MessageNotification(message, severity)
        props.displaySnackBar(notification);

        if ('access' in result) {
            props.handleClose();
        }
    }
    return (
        <Button
            variant="contained"
            sx={ButtonDialogSX}
            key="login-dialog-signIn"
            style={{
                minHeight: "40px",
                maxHeight: "40px"
            }}
            onClick={() => {
                handleLoginButton(props)
                    .then(() => {
                        props.handleClose();
                    })
                    .catch(error => {
                        if ('status' in error && error.status === "FETCH_ERROR") {
                            props.displaySnackBar(
                                new MessageNotification(
                                    <div style={{
                                        verticalAlign: "middle"
                                    }}>
                                        {localizedStrings.FETCH_ERROR}
                                    </div>,
                                    "error")
                            )
                        }
                    });
            }}
            startIcon={<VpnKey />}
        >
            {localizedStrings.LOGIN_LABEL}
        </Button>
    )
}
