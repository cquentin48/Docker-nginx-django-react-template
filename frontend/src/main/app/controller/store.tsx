import { configureStore } from "@reduxjs/toolkit";
import { api } from "./user/userSlice";

export const store = configureStore({
    reducer: {
        [api.reducerPath]: api.reducer
    },
    middleware: getDefaultMiddleware =>
        getDefaultMiddleware().concat(api.middleware)
})
