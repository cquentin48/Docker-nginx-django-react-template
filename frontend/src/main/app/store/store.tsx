import { configureStore } from "@reduxjs/toolkit";
import { api } from "./user/userSlice";

/*
import UserSlice, { type UserSliceState } from "./user/userSlice";
import { type ToolkitStore } from "@reduxjs/toolkit/dist/configureStore";

export const store = configureStore({
    reducer: {
        user: UserSlice
    },
    middleware: (getDefaultMiddleWare) =>
        getDefaultMiddleWare().concat()
}
);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type Store = ToolkitStore<
{
    user: UserSliceState
},
AnyAction,
[
    ThunkMiddleware<
    {
        user: UserSliceState
    },
    AnyAction
    >
]
>;
*/

export const store = configureStore({
    reducer: {
        [api.reducerPath]: api.reducer
    },
    middleware: getDefaultMiddleware =>
        getDefaultMiddleware().concat(api.middleware)
})
