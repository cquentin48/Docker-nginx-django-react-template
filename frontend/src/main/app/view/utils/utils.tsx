/**
 * Check if the web application is not running inside a docker container
 * - `true` yes
 * - `false` no
 * @returns {boolean} whether the app is inside a docker container or not
 */
export function isNotInContainer (): boolean {
    return process.env.REACT_APP_IS_IN_DOCKER_COMPOSE_MODE !== "1";
}
