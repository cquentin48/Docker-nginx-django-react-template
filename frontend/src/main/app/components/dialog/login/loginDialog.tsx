import React from "react";

import {
    Button,
    Container,
    Dialog,
    DialogContentText,
    DialogTitle,
    TextField
} from "@mui/material";

import { CleaningServices } from "@mui/icons-material";
import { LoginButton } from "./loginButton";
import { ButtonDialogSX, TextFieldDialogSX } from "../styles";
import localizedStrings from "../../../model/locale/locale";
import type MessageNotification from "../../../model/message/message";

type AlertColor = "success" | "info" | "warning" | "error";

interface LoginDialogProps {
    open: boolean
    handleClose: () => void
    displayNotificationFunction: (message: MessageNotification) => void
}

interface LoginDialogState {
    username: string
    password: string
    snackbarMessage: string
    snackbarOpened: boolean
    snackbarSeverity: AlertColor
}

class LoginDialog extends React.Component<LoginDialogProps, LoginDialogState> {
    constructor (props: LoginDialogProps) {
        super(props);
        this.state = {
            username: "username",
            password: "password",
            snackbarMessage: "",
            snackbarOpened: false,
            snackbarSeverity: "info"
        };

        this.resetInputValues = this.resetInputValues.bind(this);
        this.updatePasswordInput = this.updatePasswordInput.bind(this);
        this.updateUsernameInput = this.updateUsernameInput.bind(this);
    }

    resetInputValues (): void {
        this.setState({
            username: "username",
            password: "password"
        });
    }

    /**
   * Update the username into the class state
   * @param usernameUpdateEvent username value change event
   */
    updateUsernameInput (
        usernameUpdateEvent: React.ChangeEvent<HTMLInputElement>
    ): void {
        this.setState({
            username: usernameUpdateEvent.target.value
        });
    }

    /**
   * Check if the password and the username
   * have been set
   */
    areTheInputsSet (): boolean {
        const state = this.state;
        return state.username !== "" && state.password !== "";
    }

    /**
   * Update the password into the class state
   * @param passwordChangeEvent password value change event
   */
    updatePasswordInput (
        passwordChangeEvent: React.ChangeEvent<HTMLInputElement>
    ): void {
        this.setState({
            password: passwordChangeEvent.target.value
        });
    }

    displaySnackBar (snackbarMessage: string): void {
        this.setState({
            snackbarMessage
        });
    }

    render (): JSX.Element {
        const props = this.props;
        const state = this.state;

        return (
            <div>
                <Dialog
                    onClose={props.handleClose}
                    open={props.open}
                    fullWidth
                    maxWidth="md"
                >
                    <DialogTitle
                        sx={{
                            textAlign: "center"
                        }}
                    >
            Connexion
                    </DialogTitle>
                    <TextField
                        label="Login"
                        defaultValue="Login"
                        sx={{
                            margin: "16px 24px 16px"
                        }}
                        onChange={this.updateUsernameInput}
                    />
                    <TextField
                        label="Password"
                        defaultValue="Password"
                        type="password"
                        sx={TextFieldDialogSX}
                        onChange={this.updatePasswordInput}
                    />
                    <Container maxWidth={false}>
                        <LoginButton
                            username={state.username}
                            password={state.password}
                            displaySnackBar={props.displayNotificationFunction}
                            handleClose={props.handleClose}
                        />
                        <Button
                            variant="outlined"
                            sx={ButtonDialogSX}
                            key="login-dialog-reset"
                            style={{
                                minHeight: "40px",
                                maxHeight: "40px"
                            }}
                            onClick={() => {
                                this.resetInputValues();
                            }}
                            startIcon={<CleaningServices />}
                        >
                            {localizedStrings.RESET_INPUT_LABEL}
                        </Button>
                        <DialogContentText>
                            {localizedStrings.LOST_PASSWORDS_TEXT}
                        </DialogContentText>
                    </Container>
                </Dialog>
            </div>
        );
    }
}

export default LoginDialog;
