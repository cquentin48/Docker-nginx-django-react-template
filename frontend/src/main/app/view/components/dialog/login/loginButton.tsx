import React from "react";
import { Button } from "@mui/material";
import { VpnKey } from "@mui/icons-material";
import { type VariantType } from "notistack";
import { ButtonDialogSX } from "../styles";

import localizedStrings from "../../../../app/locale/locale";
import { useLoginMutation } from "../../../../controller/user/userSlice";
import { type APIResponse } from "../../../../model/user/httpRequestInterfaces";
import UserFactory from "../../../../model/user/userFactory";
import { isNotInContainer } from "../../../utils/utils";
import NotificationBuilder from "../../../../app/notification/notificationBuilder";

interface LoginButtonProps {
    username: string
    password: string
    handleClose: () => void
}

export const LoginButton = (props: LoginButtonProps): JSX.Element => {
    /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
    const [login, { isLoading }] = useLoginMutation();

    const handleLoginButton = async (props: LoginButtonProps): Promise<void> => {
        const loginInput = { username: props.username, password: props.password }
        const result = await login(loginInput).unwrap()

        let message;
        let severity: VariantType;
        if ('detail' in result) {
            message = localizedStrings.FAILED_LOGIN
            severity = "error"
        } else {
            const data: APIResponse = result;
            UserFactory.updateUser(data.access);
            message = localizedStrings.SUCCESSFUL_LOGIN
            severity = "success"
        }

        NotificationBuilder.displayNotification(message, severity);

        /* istanbul ignore next */
        if ('access' in result) {
            props.handleClose();
        }
    }

    /* istanbul ignore next */
    return (
        <Button
            variant="contained"
            sx={ButtonDialogSX}
            key="login-dialog-signIn"
            style={{
                minHeight: "40px",
                maxHeight: "40px"
            }}
            onClick={() => {
                if (isNotInContainer()) {
                    UserFactory.updateUser("");
                    props.handleClose();
                } else {
                    handleLoginButton(props)
                        .then(() => {
                            props.handleClose();
                        })
                        .catch(error => {
                            if ('status' in error && error.status === "FETCH_ERROR") {
                                NotificationBuilder.displayNotification(localizedStrings.FETCH_ERROR, "error")
                            }
                        });
                }
            }}
            startIcon={<VpnKey />}
        >
            {localizedStrings.LOGIN_LABEL}
        </Button>
    )
}
