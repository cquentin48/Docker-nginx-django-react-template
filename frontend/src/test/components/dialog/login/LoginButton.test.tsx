import React from "react"

import { testIgnoredInNoContainer } from "../../../utils"
import { LoginButton } from "../../../../main/app/view/components/dialog/login/loginButton"
import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import { store } from "../../../../main/app/controller/store";
import axios from "axios";

test("Is login button correctly rendered", () => {
    // Given
    const inputs = {
        username: "myUser",
        password: "password"
    }
    const handleCloseDialogFunction = (): void => {};

    // Acts
    const tree = renderer.create(
        <Provider store={store}>
            <LoginButton
                handleClose={handleCloseDialogFunction}
                username={inputs.username}
                password={inputs.password}
            />
        </Provider>
    ).toJSON();

    // Asserts
    expect(tree).toMatchSnapshot();
});

testIgnoredInNoContainer("Login button with correct inputs", () => {
    // Before
    axios.post("http://0.0.0.0/api/v1/user/register", {
        username: "user",
        password: "password",
        email: "mail@mail.com",
        first_name: "F.",
        last_name: "L."
    }).then(result => {
        // Given
        const inputs = {
            username: "myUser",
            password: "password"
        }
        const handleCloseDialogFunction = (): void => {};

        // Acts
        render(
            <Provider store={store}>
                <LoginButton
                    handleClose={handleCloseDialogFunction}
                    password={inputs.password}
                    username={inputs.username}
                />
            </Provider>
        )
        const button = screen.getByRole('button')
        fireEvent.click(button)

        // Asserts
        expect(localStorage.getItem("user")).not.toBe(null)
        expect(localStorage.getItem("user")).not.toBe(undefined)
    }).catch(_ => {
        fail("The login button should update the user!")
    })
})
