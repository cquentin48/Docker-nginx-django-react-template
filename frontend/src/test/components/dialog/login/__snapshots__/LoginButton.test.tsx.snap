// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Is login button correctly rendered 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-nwv6f9-MuiButtonBase-root-MuiButton-root"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "maxHeight": "40px",
      "minHeight": "40px",
    }
  }
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="VpnKeyIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12.65 10C11.83 7.67 9.61 6 7 6c-3.31 0-6 2.69-6 6s2.69 6 6 6c2.61 0 4.83-1.67 5.65-4H17v4h4v-4h2v-4H12.65zM7 14c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2z"
      />
    </svg>
  </span>
  Login
</button>
`;
